### DISCUSSION TITLE
Derive macro for generating method tuples from structs

### DISCUSSION THREAD

**Discussion by eaglesemanation** [diwic/dbus-rs#456] [Ideas]
I've been working on a small personal project, which is my first experience with both DBus and dbus-rs. To be more exact, I wanted to interact with [Mutter DisplayConfig](https://gitlab.gnome.org/GNOME/mutter/-/blob/main/data/dbus-interfaces/org.gnome.Mutter.DisplayConfig.xml), and almost immediately I came to a conclusion that managing needed arguments in form of tuples would be a nightmare. 
Instead, I spent 3 days writing a proof of concept derive macro, which takes a struct and generates a tuple type from it, plus implements TryFrom in both directions. Here is [my derive macro](https://github.com/eaglesemanation/gnome-randr/blob/main/dbus-derive/src/lib.rs#L33) and here is an [example of how I'm using it](https://github.com/eaglesemanation/gnome-randr/blob/main/randr/src/display_config.rs#L35).
So, my question: Is this something that is worth trying to upstream? I'd love to get any kind of feedback on this.

**Comment by diwic** [comment]
Hey, interesting stuff! :-)

As for trying to upstream it - let's have a discussion about it and see how things go? 

I had a bird's eye look at the code, and so my main question is about the traits (DbusArg and DbusDictKey), any reason you can't use existing traits such as `AppendAll`, `ArgAll` and `ReadAll`? If you derive those instead, you shouldn't have to convert structs back and forth to a tuple just to be able to send them over dbus.

As for this question:
```
// I have no clue how dbus-rs has f64 and std::fs::File as a DictKey, they don't implement
// core::hash::Hash
```
The reason is that since dbus protocol specifies that f64 and Files can be dict keys I do my best to support that. But it would be rather weird and I haven't seen anyone using it in practice.

**Reply by eaglesemanation** [reply]
Thanks for taking a look.
1. I need a custom trait for mapping types that are not 1:1, for example in my use case I'm defining a `enum Transform` which maps to `u8` as Dbus argument. I definitely want end user to be able to implement that trait for whatever crazy data type they want. This in turn makes containers implementation not trivial, so for example, I have to iterate through elements in `Vec<Transform>` and apply DbusArg conversion to `u8` for every element. Originally I tried implementing `DbusArg<T> where T: dbus::arg::Arg`, but because there are no specializations in Rust - I couldn't do custom logic for Vec. I might be overthinking this, let me know if you see any better approach.
2. I originally tried to add both f64 and File, but DbusArg<HashMap<...>> implementation would not compile screaming at me, because they don't implement core::hash::Hash. I wasn't exactly sure how this codebase compiles without the same issues, so that comment was mostly a reminder for myself to dig into this.

**Reply by eaglesemanation** [reply]
Actually, nevermind, I thought through what you're actually suggesting, and I'll try implementing AppendAll and such directly. Got stuck in a mindset of specific solution, will try yours instead, will give an update probably this week

**Reply by diwic** [reply]
Right, so there is already code for generating some or maybe all of this in the dbus-codegen crate, just that it starts with an xml file instead of Rust types. I've not written any proc-macro crates so I'm not that familiar with syn, quote and friends, but any chance we could share code between the two code generation methods?

**Reply by eaglesemanation** [reply]
I'm almost done with a rewrite that uses your traits instead of custom ones. I don't know how much code can be reused from your codegen crate, because it seems to be mostly relying on String templating, while mine is first parsed from TokenStream into AST and then back. Making some sort of adapters between those 2 interfaces might be more code than proc-macro itself, but I'll let you judge that when my rewrite is done.

**Reply by eaglesemanation** [reply]
I'm done with the rewrite, and except 1 small bug annotated with `//FIXME`, my poor man's version of randr works as it did before. derive crate is almost entirely rewritten, I tried to replicate how this library uses `macro_rules!` for implementing traits on tuples, a lot of it is just copy-paste with `quote!` syntax adjustments. Hope it makes it more readable for you.

**Comment by diwic** [comment]
Hi! I've been looking at your crate and it looks better now :-)

I think I understand most of it, it might need a bit more of tests/documentation/readme so users understand how to use it. Also the "rename" option of DbusPropMap seems to be untested and undocumented.

I have a question about DbusArgs - it seems to assume the args is in a struct so it should probably be named DbusStruct instead? A version that does not descend into a struct might be even more useful - you could e g use it with `Message::read_all` or `Message::append_all`. Not saying you have to write one, but that is what I would expect a DbusArgs to do.

**Reply by eaglesemanation** [reply]
Hey, sorry for being silent, end of quarter got a bit noisy. I'm not writing documentation yet because I want to get to the point where I can use it myself, to see if UX makes sense. But I definitely want to add more tests, especially for compilation error messages, so it's actually readable.

As for DbusArgs - right now it also allows to derive on `struct WrapperType(u32);` and that will have a signature of `u` instead of `(u)`. I don't have a use case for this behavior right now, and your suggestion makes more sense, I'll adjust it.

**Reply by eaglesemanation** [reply]
Added a bunch more tests, tried my best to make all the errors readable. Unfortunately, error for type inside a struct not implementing Arg + Append + Get is really verbose, but I don't think I can do anything to make it better. Best thing that I came up with - add documentation saying "if you see this huge error - you forgot to derive(DbusStruct) for nested type".

Haven't started on documentation yet, will do that after I finally finish with original project.

As for your suggestion of implementation that does not descend into a struct - I'm really struggling trying to understand what you meant by that, could you give some pseudo code / example of how that would look like? I was mostly referencing implementation for tuples, and they seem to use iter.recurse() the same way I do.

**Reply by diwic** [reply]
I had another look at your code. Maybe what's confusing you is that tuples were originally used for structs only, but now I also use them in some cases for appending multiple arguments to a message. 

E g, an integer and a string together would be:

1) Used inside a struct with signature `(us)`. That struct is a single argument and would then implement Arg, Append and Get. Let's call this DbusStruct.

2) Used as two separate arguments. E g, you might receive a signal with one integer and one string argument. The signal's signature is then `us`. Those arguments (together) would then implement ArgAll, AppendAll and ReadAll. Let's call this DbusArgs. And ReadAll would just read an integer and then a string, without any recursing. 

But there is rarely a scenario where the same struct would be used for both 1) and 2). So I don't think the same derive should implement both Arg and ArgAll, this would typically not make sense, split it up instead.

Does that make sense to you?

**Reply by eaglesemanation** [reply]
Makes a lot of sense, I'll keep Arg + Append + Get as DbusStruct and split ArgAll + AppendAll + ReadAll into DbusArgsBundle or something like that.

Out of curiosity - why Arg, Append and Get are separate traits instead of single one? I've been thinking about how I could make compilation error for missing trait implementation more readable, and I think maybe having 1 trait would make it simpler to follow. I'll try doing a wrapper trait and a blanket implementation to see if that's true.

**Reply by diwic** [reply]
It is for maximum flexibility and performance, as some interesting Rust types might be appendable but not get-able, or the other way around. E g `&[&str]` might be appendable but not get-able.

Btw, "Bundle" is not Dbus terminology so I'd prefer `DbusArgs`, or maybe just `Args` because "dbus" is in the module name.

**Reply by eaglesemanation** [reply]
### Doc test

Right now I'm working on documentation, wanted to add an example and got a really weird issue - linking libsystemd.so.0 fails while compiling doc test. Here is exact error:
```
running 1 test
test dbus-derive/src/lib.rs - derive_dbus_propmap (line 84) ... FAILED

failures:

---- dbus-derive/src/lib.rs - derive_dbus_propmap (line 84) stdout ----
Test executable failed (exit status: 127).

stderr:
/tmp/rustdoctestBfG30s/rust_out: error while loading shared libraries: libsystemd.so.0: cannot open shared object file: No such file or directory



failures:
    dbus-derive/src/lib.rs - derive_dbus_propmap (line 84)
```
Considering that my unit tests work, I think I don't have any issues with having library on my system. For now I'll just skip testing examples.

### Flatten

Another observation - I wanted to add a feature to flatten a struct that implements `ArgAll` into another struct. So for example:
```rs
#[derive(DbusArgs)]
struct Flattened {
	x: i32,
	y: i32,
}

#[derive(DbusStruct)]
struct Wrapper {
	#[dbus_struct(flatten)]
	flat: Flattened
}

assert_eq!(Wrapper::signature(), "(ii)")
```
The issue is that I cannot really define `type strs` in `ArgAll`, because I cannot figure out the amount of fields inside `Flattened` struct at the moment of macro expansion. Assuming that we could break compatibility - is there reason why `type strs` needs to exist? Couldn't this work:
```patch
pub trait ArgAll {
-   /// A tuple of &static str. Used for introspection.
-   #[allow(non_camel_case_types)] // Note: This should be changed for 0.9 - but for now, don't break backwards compatibility
-   type strs;
    /// Enumerates all arguments with their signatures (introspection helper method).
-   fn strs_sig<F: FnMut(&'static str, Signature<'static>)>(a: Self::strs, f: F);
+   fn strs_sig<F: FnMut(&'static str, Signature<'static>)>(a: &[&'static str>], f: F);
}
```

**Reply by eaglesemanation** [reply]
I've added documentation for each derive macro, if it looks good to you - I'll start working on a PR for this repo

**Reply by diwic** [reply]
Cool! 

I'm not sure about the question about removing `type strs`. Maybe start upstreaming what you've got so far, then if you like you could try removing it (and work on the flatten feature) and see if you run into trouble somewhere?

We've had 0.9 for a long time so I'm open to releasing something backwards incompatible, especially as this seems like an easy change to adapt to for users.

**Reply by diwic** [reply]
As for the doc test question, I do want `cargo test` to work before merging. Not sure if the problem is with your environment or with the example, we'll just have to see I guess.

**Reply by eaglesemanation** [reply]
Seems like linkage issue with rustdoc was just recently resolved: https://github.com/rust-lang/cargo/pull/13490
That issue might be unique to my setup because I'm on Fedora with Nix for dev dependencies, so dynamic linkage is a bit of a mess. Running tests with `--features vendored` worked perfectly fine.
Made a draft PR here: https://github.com/diwic/dbus-rs/pull/459.
Want to make DbusEnum more general purpose, not just enum specific, so maybe DbusType, or DbusInto. That might be useful for some wrapper types, or maybe tightly packed values. After that will move it out of draft.
